Fix for raw inflate with directory

Xubor opened the pull request:
Fixes https://github.com/nodeca/pako/issues/155
`inflateSetDictionary()` is called at different conditions during normal and raw modes, according to http://zlib.net/manual.html#Advanced 
Also providing basic tests for raw inflate with directory, and fix for broken test Inflate with dictionary/spdy dictionary

puzrin commented on test/inflate.js:
```diff
@@ -183,8 +183,18 @@ describe('Inflate with dictionary', function () {
 
-    testInflate(samples, { dictionary: spdyDict }, { dictionary: helpers.spdyDict });
+  it('should throw if directory is not supplied to raw inflate', function () {
+    var dict = new Buffer('abcdefghijklmnoprstuvwxyz');
```
`new Buffer()` -> `Buffer.from()`

puzrin replied:
Or may be it's possible to not use Buffer at all, casting string to Uint8Array.

Xubor replied:
No problem. I have just used what was in previous test.

Xubor replied:
Resolved in https://github.com/nodeca/pako/pull/156/commits/cfc92c2fb860f9c8923fc938c970fd8727cc7e84

puzrin commented on lib/inflate.js:
```diff
@@ -144,6 +144,17 @@ function Inflate(options) {
+  if (opt.dictionary) {
+    // Convert data if needed
+    if (typeof opt.dictionary === 'string') {
+      opt.dictionary = strings.string2buf(opt.dictionary);
```
It's not correct to modify options object.

puzrin replied:
Ah, sorry, that's a cloned object. Then hacking type is acceptable, i think.

puzrin commented on test/inflate.js:
```diff
@@ -183,8 +183,18 @@ describe('Inflate with dictionary', function () {
 
   it('spdy dictionary', function () {
     var spdyDict = require('fs').readFileSync(require('path').join(__dirname, 'fixtures', 'spdy_dict.txt'));
+    testInflate(samples, { dictionary: spdyDict }, { dictionary: spdyDict });
```
ups... thanks :)

puzrin commented on test/inflate.js:
```diff
@@ -183,8 +183,18 @@ describe('Inflate with dictionary', function () {
+  it('tests raw inflate with spdy dictionary', function () {
+    var spdyDict = require('fs').readFileSync(require('path').join(__dirname, 'fixtures', 'spdy_dict.txt'));
+    testInflate(samples, { raw: true, dictionary: spdyDict }, { raw: true, dictionary: spdyDict });
+  });
```
Could you add tests with all supported dictionary types (array, typed array)? Pako has very high coverage, and it would be nice to keep it high.

Use [`make cover`](https://github.com/nodeca/pako/blob/master/Makefile#L30) to check

Xubor replied:
That should by another PR as it is not related to bug in raw inflate.

Xubor replied:
Added sufficient tests in https://github.com/nodeca/pako/pull/156/commits/b4f1c205efdf998f512cbf5484445fc0b30f8051

puzrin commented:
LGTM. See minor comments about tests. Since this part of code is touched, it would be nice to clean related things. But if you a busy, i can merge as is and do the rest myself. Let me know.

Xubor commented:
In susequent commits I have removed the `Buffer` usage from the relevant test section and added basic tests sufficient to cover all paths of touched code in `Inflate()`. 
But beware, that does not guarantee that code for the dictionary format conversion is complete and sound, especially for browser environments. This code was present in the source before, without appropriate tests.

Xubor commented:
@puzrin ping?

Xubor commented:
Thank you, I appreciate your good work!
As this fixes real bug, I suggest issuing new version.

puzrin commented:
Published new release with your fixes. Thanks for your help!
